### PR TITLE
[SPARK-13436][SPARKR] Added parameter drop to subsetting operator [

### PR DESCRIPTION
**Pull request by olarayej**
Added parameter drop to subsetting operator [. This is useful to get a Column from a DataFrame, given its name. R supports it. 

In R:

```
> name <- "Sepal_Length"
> class(iris[, name])
[1] "numeric"
```

Currently, in SparkR:

```
> name <- "Sepal_Length"
> class(irisDF[, name])
[1] "DataFrame"
```

Previous code returns a DataFrame, which is inconsistent with R's behavior. SparkR should return a Column instead. Currently, in order for the user to return a Column given a column name as a character variable would be through `eval(parse(x))`, where x is the string `"irisDF$Sepal_Length"`. That itself is pretty hacky. `SparkR:::getColumn()`is another choice, but I don't see why this method should be externalized. Instead, following R's way to do things, the proposed implementation allows this:

```
> name <- "Sepal_Length"
> class(irisDF[, name, drop=T])
[1] "Column"

> class(irisDF[, name, drop=F])
[1] "DataFrame"
```

This is consistent with R:

```
> name <- "Sepal_Length"
> class(iris[, name])
[1] "numeric"
> class(iris[, name, drop=F])
[1] "data.frame"
```

### CODE REVIEW

**Comment by gatorsmile** [comment]
Please update your PR title. Thanks!

**Comment by olarayej** [comment]
@felixcheung @sun-rui @shivaram Can you folks please take a look at this one? Thank you!

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -1217,29 +1217,38 @@ setMethod("[[", signature(x = "DataFrame", i = "numericOrcharacter"),
 
 #' @rdname subset
```
Comment (shivaram) [inline R/pkg/R/DataFrame.R]: can you add some documentation on what `drop` does in this context ?

File: R/pkg/R/DataFrame.R [hunk 2]
```diff
@@ -1217,29 +1217,38 @@ setMethod("[[", signature(x = "DataFrame", i = "numericOrcharacter"),
...
-          function(x, i, j, ...) {
-            if (is.numeric(j)) {
-              cols <- columns(x)
-              j <- cols[j]
-            }
-            if (length(j) > 1) {
-              j <- as.list(j)
+setMethod("[", signature(x = "DataFrame"),
+          function(x, i, j, ..., drop=T) {
+            # Perform filtering first if needed
```
Comment (shivaram) [inline R/pkg/R/DataFrame.R]: Also I think `drop=F` should be default if we want to maintain backwards compatibility.

File: R/pkg/R/DataFrame.R [hunk 3]
```diff
@@ -1217,29 +1217,38 @@ setMethod("[[", signature(x = "DataFrame", i = "numericOrcharacter"),
...
+            filtered <- if (missing(i)) {
+              x
+            } else {
+              if (class(i) != "Column") {
+                stop(paste0("Expressions other than filtering predicates are not supported ",
+                      "in the first parameter of subsetting operator [."))
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: suggest `extract operator [ or subset()`, as [ is referred to as 'extract' [here](https://stat.ethz.ch/R-manual/R-devel/library/base/html/Extract.html), and subset() calls also go here

**Comment by felixcheung** [comment]
Hmm, just as #11336 wouldn't it be confusing then to return a Column, particularly since Column in SparkR is pretty restricted, unlike in R?

**Comment by sun-rui** [comment]
https://stat.ethz.ch/R-manual/R-devel/library/base/html/Extract.data.frame.html here for reference

File: R/pkg/R/DataFrame.R (hunk 2 above)
Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: Maybe for Spark 2.0 we can sacrifice some backwards compatibility and have the R user convenience as long term benefit?

File: R/pkg/R/DataFrame.R [hunk 4]
```diff
@@ -1271,12 +1280,8 @@ setMethod("[", signature(x = "DataFrame", i = "Column"),
 #'   subset(df, select = c(1,2))
 #' }
 setMethod("subset", signature(x = "DataFrame"),
-          function(x, subset, select, ...) {
-            if (missing(subset)) {
-              x[, select, ...]
-            } else {
-              x[subset, select, ...]
-            }
+          function(x, subset, select, drop=T, ...) {
```
Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: inconsistent with drop = FALSE in the signature of subset() in the R base package?

File: R/pkg/R/DataFrame.R [hunk 5]
```diff
@@ -1217,29 +1217,38 @@ setMethod("[[", signature(x = "DataFrame", i = "numericOrcharacter"),
...
-setMethod("[", signature(x = "DataFrame", i = "missing"),
-          function(x, i, j, ...) {
-            if (is.numeric(j)) {
-              cols <- columns(x)
-              j <- cols[j]
-            }
-            if (length(j) > 1) {
-              j <- as.list(j)
+setMethod("[", signature(x = "DataFrame"),
+          function(x, i, j, ..., drop=T) {
```
Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: coding style:
  drop = T

Many other places have such coding style issue too

File: R/pkg/R/DataFrame.R (hunk 1 above)
Comment (olarayej) [inline R/pkg/R/DataFrame.R]: Done!

**Comment by olarayej** [comment]
@felixcheung @shivaram @sun-rui I have addressed all your comments. Do we have a consensus on the default value for drop? I'd say drop=T makes sense cuz R does it that way. Anyways, it's just a one-line change for me. Please let me know! Thanks!

**Comment by sun-rui** [comment]
@olarayej, there are still similar coding style issues:)

**Comment by olarayej** [comment]
@sun-rui Done with the style issues. Thanks!

**Comment by olarayej** [comment]
@sun-rui @shivaram Shall we merge this?

**Comment by olarayej** [comment]
@shivaram @sun-rui @felixcheung This one's ready. Shall we merge?

**Comment by shivaram** [comment]
Thanks @olarayej for updating this. I think the code looks good. The only thing I'm not sure of is the default value for `[` - I see that you have gone through SparkR's internal code base to add `drop = F`. The reason I prefer keep `drop=F` as the default is that I'm not sure if we have missed some place where we use `[`. While I see the argument that its the default in normal R, I feel like it might not be worth changing the default if we'll run into other unexpected bugs.

**Comment by olarayej** [comment]
@shivaram I've changed default value to drop=F as you suggested. Thanks!

**Comment by shivaram** [comment]
LGTM. Merging this. Thanks @olarayej

**Pull request closed without merging**